# Phase 3: Control Tower Telegramé€£æº - å®Œäº†å ±å‘Š

**å®Œäº†æ—¥æ™‚**: 2026-02-04
**ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹**: âœ… å®Œäº†

---

## æ¦‚è¦

Phase 3ã§ã¯ã€Control Tower ã®ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ã‚’ Telegram ã®ãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã«è¡¨ç¤ºã—ã€ãƒªã‚¢ãƒ«ã‚¿ã‚¤ãƒ ã§æ›´æ–°ã™ã‚‹æ©Ÿèƒ½ã‚’å®Ÿè£…ã—ã¾ã—ãŸã€‚

---

## å®Ÿè£…å†…å®¹

### 1. Control Tower Telegram Integration

**ãƒ•ã‚¡ã‚¤ãƒ«**: `src/utils/control-tower-telegram.ts`

Telegramé€£æºã®ä¸»è¦æ©Ÿèƒ½:

```typescript
// Pinned status message management
export async function ensureStatusMessage(ctx: Context): Promise<number | null>
export async function updateStatusMessage(ctx: Context, sessionId: string): Promise<void>
export async function deleteStatusMessage(ctx: Context): Promise<void>
export async function initControlTower(ctx: Context): Promise<void>
```

### 2. Message ID Persistence

**D1ãƒ†ãƒ¼ãƒ–ãƒ«**: `jarvis_settings`

```sql
-- Control Tower message_id storage
key: control_tower_message_{chat_id}
value: {message_id}
```

- ãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã® message_id ã‚’ D1 ã«æ°¸ç¶šåŒ–
- Bot å†èµ·å‹•å¾Œã‚‚åŒã˜ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’æ›´æ–°å¯èƒ½

### 3. Graceful Recovery

**å¾©æ—§ãƒ•ãƒ­ãƒ¼**:
1. ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸æ›´æ–°æ™‚ã« `editMessageText` ã‚’è©¦è¡Œ
2. å¤±æ•—ã—ãŸå ´åˆï¼ˆmessage not foundï¼‰ã¯æ–°ã—ã„ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’ä½œæˆ
3. æ–°ã—ã„ message_id ã‚’ D1 ã«ä¿å­˜
4. å¤ã„ message_id ã‚’ã‚¯ãƒªã‚¢

### 4. Status Formatting

**è¡¨ç¤ºå½¢å¼**:
```
ğŸ¤– **JARVIS Control Tower**

çŠ¶æ…‹: ğŸ§  æ€è€ƒä¸­
Phase: Phase 3: Telegramé€£æº
ã‚¢ã‚¯ã‚·ãƒ§ãƒ³: Creating tests

_æœ€çµ‚æ›´æ–°: 10:12:34_
```

**ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ emoji ãƒãƒƒãƒ”ãƒ³ã‚°**:
- ğŸ’¤ ã‚¢ã‚¤ãƒ‰ãƒ« (idle)
- ğŸ§  æ€è€ƒä¸­ (thinking)
- ğŸ“‹ è¨ˆç”»ä¸­ (planning)
- âš™ï¸ å®Ÿè¡Œä¸­ (executing)
- â³ æ‰¿èªå¾…ã¡ (waiting_approval)
- âœ… å®Œäº† (completed)
- âŒ ã‚¨ãƒ©ãƒ¼ (error)

---

## çµ±åˆãƒã‚¤ãƒ³ãƒˆ

### 1. control-tower-helper.ts ã®ä¿®æ­£

`updateStatus()`, `startPhase()`, `completePhase()` ã« Context ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’è¿½åŠ :

```typescript
export function updateStatus(
  sessionId: string,
  statusType: string,
  phase?: string | null,
  action?: string | null,
  ctx?: Context | null  // â† è¿½åŠ 
): void {
  // ... D1 update

  // Telegramæ›´æ–°ï¼ˆéåŒæœŸã€ã‚¨ãƒ©ãƒ¼ç„¡è¦–ï¼‰
  if (ctx) {
    getTelegramIntegration()
      .then((integration) => integration.updateStatusMessage(ctx, sessionId))
      .catch((error) => {
        console.error('[ControlTower] Failed to update Telegram message:', error);
      });
  }
}
```

### 2. streaming.ts ã®ä¿®æ­£

ã™ã¹ã¦ã® `updateStatus()` å‘¼ã³å‡ºã—ã« `ctx` ã‚’è¿½åŠ :

```typescript
if (sessionId) {
  updateStatus(sessionId, "thinking", null, preview, ctx);  // â† ctxè¿½åŠ 
}
```

### 3. notification-buffer.ts ã®ä¿®æ­£

`startPhaseDB()` ã¨ `completePhaseDB()` å‘¼ã³å‡ºã—ã« `ctx` ã‚’è¿½åŠ :

```typescript
if (sessionId) {
  startPhaseDB(sessionId, phaseName, ctx);  // â† ctxè¿½åŠ 
}
```

### 4. Lazy Loading

å¾ªç’°ä¾å­˜å›é¿ã®ãŸã‚ã€Telegramçµ±åˆã‚’é…å»¶ãƒ­ãƒ¼ãƒ‰:

```typescript
let telegramIntegration: any = null;
async function getTelegramIntegration() {
  if (!telegramIntegration) {
    telegramIntegration = await import('./control-tower-telegram');
  }
  return telegramIntegration;
}
```

---

## Command Handler

### /tower ã‚³ãƒãƒ³ãƒ‰

**ãƒ•ã‚¡ã‚¤ãƒ«**: `src/handlers/tower.ts`

```typescript
export async function handleTower(ctx: Context): Promise<void> {
  await initControlTower(ctx);
}
```

**ä½¿ã„æ–¹**:
```
/tower
```

**å®Ÿè¡Œçµæœ**:
1. ãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’ä½œæˆ
2. ç¾åœ¨ã®ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ã‚’è¡¨ç¤º
3. ã€Œâœ… Control Tower åˆæœŸåŒ–å®Œäº†ã€ã¨ç¢ºèª

---

## ãƒ†ã‚¹ãƒˆçµæœ

**ãƒ•ã‚¡ã‚¤ãƒ«**: `src/tests/phase3-telegram-integration.test.ts`

### ãƒ†ã‚¹ãƒˆé …ç›®ï¼ˆ9é …ç›®ã™ã¹ã¦æˆåŠŸï¼‰

âœ… ensureStatusMessage creates new pinned message
âœ… ensureStatusMessage reuses existing message
âœ… ensureStatusMessage recovers when message is deleted
âœ… updateStatusMessage updates message text
âœ… updateStatusMessage handles missing message gracefully
âœ… deleteStatusMessage unpins and deletes message
âœ… initControlTower creates pinned message and confirms
âœ… Multiple status updates maintain single pinned message
âœ… Status formatting includes all relevant fields

```
 9 pass
 0 fail
 28 expect() calls
```

---

## å‹•ä½œãƒ•ãƒ­ãƒ¼

### æ­£å¸¸ãƒ•ãƒ­ãƒ¼

1. ãƒ¦ãƒ¼ã‚¶ãƒ¼ãŒ `/tower` ã‚³ãƒãƒ³ãƒ‰ã‚’å®Ÿè¡Œ
2. Bot ãŒãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’ä½œæˆ
3. message_id ã‚’ D1 ã«ä¿å­˜ (`control_tower_message_{chat_id}`)
4. ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹æ›´æ–°æ™‚ã« `editMessageText` ã§ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’æ›´æ–°
5. ãƒªã‚¢ãƒ«ã‚¿ã‚¤ãƒ ã§ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ãŒåæ˜ ã•ã‚Œã‚‹

### å¾©æ—§ãƒ•ãƒ­ãƒ¼

1. ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãŒå‰Šé™¤ã•ã‚ŒãŸï¼ˆã¾ãŸã¯Botå†èµ·å‹•å¾Œã«è¦‹ã¤ã‹ã‚‰ãªã„ï¼‰
2. `editMessageText` ãŒ 400 ã‚¨ãƒ©ãƒ¼ã‚’è¿”ã™
3. æ–°ã—ã„ãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’è‡ªå‹•ä½œæˆ
4. æ–°ã—ã„ message_id ã‚’ D1 ã«ä¿å­˜
5. ä»¥é™ã¯æ–°ã—ã„ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’æ›´æ–°

---

## ã‚¨ãƒ©ãƒ¼å‡¦ç†

### 1. Message Not Found (400 Error)

```typescript
if (error.error_code === 400 && error.description?.includes('message to edit not found')) {
  const key = `control_tower_message_${chatId}`;
  controlTowerDB.deleteSetting(key);
  console.log('[ControlTowerTelegram] Cleared invalid message_id, will create new on next update');
}
```

### 2. Non-blocking Updates

- Telegramæ›´æ–°ã¯éåŒæœŸã§å®Ÿè¡Œ
- ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿã—ã¦ã‚‚ D1 è¨˜éŒ²ã¯ç¶™ç¶š
- ãƒ­ã‚°å‡ºåŠ›ã®ã¿ã§æ—¢å­˜æ©Ÿèƒ½ã«å½±éŸ¿ãªã—

### 3. Null-safe Context

```typescript
if (ctx) {
  // Telegramæ›´æ–°
}
```

---

## ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹å½±éŸ¿

- Telegram API å‘¼ã³å‡ºã—ã¯éåŒæœŸï¼ˆnon-blockingï¼‰
- ãƒ†ã‚¹ãƒˆçµæœ: 9 tests in 41msï¼ˆå¹³å‡ 4.6ms/testï¼‰
- å®Ÿç’°å¢ƒã§ã®ã‚ªãƒ¼ãƒãƒ¼ãƒ˜ãƒƒãƒ‰ã¯ç„¡è¦–ã§ãã‚‹ãƒ¬ãƒ™ãƒ«
- Message ID ãŒ D1 ã«ä¿å­˜ã•ã‚Œã¦ã„ã‚‹ãŸã‚ã€æ¯å›ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’æ¤œç´¢ã™ã‚‹å¿…è¦ãªã—

---

## ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£

### 1. Chat ID Isolation

å„ãƒãƒ£ãƒƒãƒˆã”ã¨ã«ç‹¬ç«‹ã—ãŸ message_id ã‚’ä¿å­˜:

```
control_tower_message_12345
control_tower_message_67890
```

### 2. Silent Pins

```typescript
await ctx.api.pinChatMessage(chatId, message.message_id, {
  disable_notification: true,  // ã‚µã‚¤ãƒ¬ãƒ³ãƒˆãƒ”ãƒ³
});
```

---

## ä½¿ç”¨ä¾‹

### 1. åˆæœŸåŒ–

```
ãƒ¦ãƒ¼ã‚¶ãƒ¼: /tower

Bot:
ğŸ¤– **JARVIS Control Tower**

çŠ¶æ…‹: ğŸ’¤ ã‚¢ã‚¤ãƒ‰ãƒ«

_æœ€çµ‚æ›´æ–°: --:--_

Bot: âœ… Control Tower åˆæœŸåŒ–å®Œäº†
```

### 2. ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹æ›´æ–°ï¼ˆè‡ªå‹•ï¼‰

```
[ãƒ¦ãƒ¼ã‚¶ãƒ¼ãŒã‚¿ã‚¹ã‚¯ã‚’å®Ÿè¡Œä¸­]

Bot (pinned message):
ğŸ¤– **JARVIS Control Tower**

çŠ¶æ…‹: ğŸ§  æ€è€ƒä¸­
Phase: Phase 3: Implementation
ã‚¢ã‚¯ã‚·ãƒ§ãƒ³: Analyzing code...

_æœ€çµ‚æ›´æ–°: 10:15:42_
```

### 3. Phaseå®Œäº†

```
Bot (pinned message):
ğŸ¤– **JARVIS Control Tower**

çŠ¶æ…‹: âœ… å®Œäº†
Phase: Phase 3: Implementation
ã‚¢ã‚¯ã‚·ãƒ§ãƒ³: Segment 2

_æœ€çµ‚æ›´æ–°: 10:18:22_
```

---

## æ¬¡ã®ã‚¹ãƒ†ãƒƒãƒ—ï¼ˆPhase 4 å€™è£œï¼‰

1. **WebSocket ãƒªã‚¢ãƒ«ã‚¿ã‚¤ãƒ æ›´æ–°**
   - Server-Sent Events (SSE) ã«ã‚ˆã‚‹ push é€šçŸ¥
   - ãƒ–ãƒ©ã‚¦ã‚¶ãƒ€ãƒƒã‚·ãƒ¥ãƒœãƒ¼ãƒ‰ã¸ã®ãƒªã‚¢ãƒ«ã‚¿ã‚¤ãƒ é…ä¿¡

2. **Action Trace å¯è¦–åŒ–**
   - Tool å®Ÿè¡Œå±¥æ­´ã®è¡¨ç¤º
   - Duration ã®å¯è¦–åŒ–
   - ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åˆ†æ

3. **é€šçŸ¥ã‚«ã‚¹ã‚¿ãƒã‚¤ã‚º**
   - ãƒ¦ãƒ¼ã‚¶ãƒ¼ã”ã¨ã®é€šçŸ¥è¨­å®š
   - Status ã”ã¨ã® emoji ã‚«ã‚¹ã‚¿ãƒã‚¤ã‚º
   - é€šçŸ¥é »åº¦ã®èª¿æ•´

4. **ãƒãƒ«ãƒã‚»ãƒƒã‚·ãƒ§ãƒ³å¯¾å¿œ**
   - è¤‡æ•°ã®ã‚»ãƒƒã‚·ãƒ§ãƒ³ã‚’åŒæ™‚è¡¨ç¤º
   - ã‚»ãƒƒã‚·ãƒ§ãƒ³åˆ‡ã‚Šæ›¿ãˆ
   - éå»ã®ã‚»ãƒƒã‚·ãƒ§ãƒ³å±¥æ­´

---

## ã¾ã¨ã‚

âœ… **Phase 3 å®Œäº†**

- Telegram ãƒ”ãƒ³ç•™ã‚ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã«ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ã‚’è¡¨ç¤º
- `editMessageText` ã§ãƒªã‚¢ãƒ«ã‚¿ã‚¤ãƒ æ›´æ–°
- Message ID ã‚’ D1 ã«æ°¸ç¶šåŒ–ï¼ˆBotå†èµ·å‹•å¯¾å¿œï¼‰
- ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸æ¶ˆå¤±æ™‚ã®è‡ªå‹•å¾©æ—§æ©Ÿèƒ½
- 9 ã¤ã®çµ±åˆãƒ†ã‚¹ãƒˆã™ã¹ã¦ãŒæˆåŠŸ
- æ—¢å­˜æ©Ÿèƒ½ã«å½±éŸ¿ãªã—ï¼ˆnon-blocking, error-safeï¼‰
- ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ã¸ã®å½±éŸ¿ã¯ç„¡è¦–ã§ãã‚‹ãƒ¬ãƒ™ãƒ«

**Phase 1 + Phase 2 + Phase 3** ã«ã‚ˆã‚Šã€JARVIS Control Tower ã®åŸºæœ¬æ©Ÿèƒ½ãŒå®Œæˆã—ã¾ã—ãŸã€‚

- **Phase 1**: D1 ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹åŸºç›¤
- **Phase 2**: æ—¢å­˜ã‚³ãƒ¼ãƒ‰çµ±åˆï¼ˆstreaming/notification-bufferï¼‰
- **Phase 3**: Telegram UI é€£æº

æ¬¡ã® Phase ã§ã¯ã€ã‚ˆã‚Šé«˜åº¦ãªå¯è¦–åŒ–æ©Ÿèƒ½ã‚„ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åˆ†ææ©Ÿèƒ½ã‚’å®Ÿè£…å¯èƒ½ã§ã™ã€‚
